### PR TITLE
Handle variable-length string IDs in binary files

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Moved IdsMap and IdsMapConverter to lib-eda-subsetting
* Made IdsMap handle a variable-size per entity type
* Added new metadata keys for ancestor byte allocation size
  * More to be seen on this in dumper PR

@steve-fischer-200 I added you because you have a lot of context on the dumper stuff but I know I've added you to a lot of PRs lately so you can think of yourself as optional if you want.

## Testing
Ran regression tests for one study with the new format. Once these PRs are finalized I can more widely dump and test studies.

### CODE REVIEW

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/BinaryValuesStreamer.java [hunk 1]
```diff
@@ -134,24 +136,26 @@ public Iterator<VariableValueIdPair<List<Long>>> streamAncestorIds(Entity descen
...
     return new FilteredValueIterator<>(path,
         x -> true, // Do not apply any filters.
-        ancestorsWithId,
+        converter,
         Function.identity());
   }
 
   public Iterator<Long> streamUnfilteredEntityIdIndexes(Study study, Entity entity) throws IOException {
-    ListConverter<String> converter = new ListConverter<>(new StringValueConverter(BYTES_RESERVED_FOR_ID), entity.getAncestorEntities().size() + 1);
+    List<Integer> bytesReservedForAncestors = binaryFilesManager.getBytesReservedForAncestry(study, entity);
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/reducer/BinaryValuesStreamer.java]: perhaps factor this reused code into getConverter(study, entity)?

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java [hunk 2]
```diff
@@ -114,6 +122,34 @@ public Path getDoneFile(Path directory, Operation op) {
...
   }
 
+  /**
+   * For a given entity, parse the metadata JSON file to determine the number of bytes needed to store identifiers
+   * for all the entity's ancestors.
+   * @return a list of ancestors in order of closest to the farthest ascendant.
+   */
+  public List<Integer> getBytesReservedForAncestry(Study study, Entity entity) {
+    JSONObject metajson = readMetaJsonFile(getMetaJsonFile(study, entity, Operation.READ));
+    getMetaJsonFile(study, entity, Operation.READ);
```
Comment (steve-fischer-200) [inline src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java]: why are we calling getMetaJsonFile() again?

should getBytesReservedForAncestry be memoized/cached?

**Review by steve-fischer-200 — CHANGES_REQUESTED** [review]
looks great.  a couple of minor suggestions

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java [hunk 3]
```diff
@@ -114,6 +122,34 @@ public Path getDoneFile(Path directory, Operation op) {
...
+  }
+
+  /**
+   * For a given entity, parse the metadata JSON file to determine the number of bytes needed to store identifiers
+   * for the entity.
+   * @return number of bytes reserved for entity.
+   */
+  public Integer getBytesReservedForEntity(Study study, Entity entity) {
+    JSONObject metajson = readMetaJsonFile(getMetaJsonFile(study, entity, Operation.READ));
+    getMetaJsonFile(study, entity, Operation.READ);
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java]: Another double call

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMap.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * map of an entity's idIndex to its entity_id and its ancestor's entity_ids
+ * @author sfischer
+ *
+ */
+public class IdsMap extends VariableValueIdPair<List<String>> {
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMap.java]: This name implies a collection of mappings a la java.util.Map.  The Javadoc above seems to confirm this.  A better name might be RecordIdValues, IdsMapEntry, or IdMapping (in my order of preference) :)

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMapConverter.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+import org.veupathdb.service.eda.ss.model.Entity;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class IdsMapConverter implements BinarySerializer<IdsMap>, BinaryDeserializer<IdsMap> {
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMapConverter.java]: Similar naming issue as above.  Could be RecordIdValuesConverter

**Review by ryanrdoherty — APPROVED** [review]

File: src/main/java/org/veupathdb/service/eda/ss/model/reducer/BinaryValuesStreamer.java (hunk 1 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/reducer/BinaryValuesStreamer.java]: Sounds good, added this.

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java (hunk 2 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java]: Ah, that's not supposed to be there, fixed. 

I could add a cache, though it's called relatively infrequently and the cache would need to be keyed on study and entity so it's a fair bit of complexity for what it gets us. I'm open to doing it though, what are your thoughts?

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMap.java (hunk 4 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/IdsMap.java]: I agree, renamed in latest

File: src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java (hunk 3 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/variable/binary/BinaryFilesManager.java]: Thanks, fixed

**Pull request merged**